src/passes/Print.cpp

```diff
@@ -2219,15 +2219,28 @@ struct PrintExpressionContents
         printMedium(o, "br_on_cast ");
         curr->name.print(o);
         o << ' ';
-        printType(curr->ref->type);
+        if (curr->ref->type == Type::unreachable) {
+          // Need to print some reference type in the correct hierarchy rather
+          // than unreachable, and the bottom type is valid in the most
+          // contexts.
+          printType(
+            Type(curr->castType.getHeapType().getBottom(), NonNullable));
+        } else {
+          printType(curr->ref->type);
+        }
```

Comment:
Would it be simpler to use `maybePrintUnreachableReplacement` like we do elsewhere?

Reply:
I think that's a bigger hammer than we need here since we can still come up with something reasonable to print in this case.

Reply:
And to perhaps add to that, I do think it's useful to avoid `maybePrintUnreachableReplacement` when possible because printing (some version of) the real expression gets us better round-tripping through text, which can be super useful for reducing test cases and general debugging.

Reply:
Got it, good points.